### PR TITLE
Documenting the extension model of the Delivery API

### PR DESCRIPTION
**Pull request by elit0451**
## Details
The Delivery API has been designed with extensibility in mind, so these new articles document how you can customize the API response based on your needs.

- Other updates related to:
  - Adding a Rebuilding the _DeliveryApiContentIndex_ limitation
  - And fixing some params spacing in the Getting started with the Delivery API article
  - Updating the Delivery API response examples to include the new `createDate` and `updateDate` fields, introduced by https://github.com/umbraco/Umbraco-CMS/pull/14427

## Product & version
Umbraco CMS v12.

------
## For Docs team
The "output expansion" link in the Custom property editors support article seems to be broken. ❗ Help needed 🙏 

Links to "Creating a Property Editor" and "Property Value Converters" articles are broken - check the comment below. ❗ Help needed 🙏

_❗❗❗ Please be aware that when making changes through the GitBook platform, some weird spacing might show up in the Content Delivery API article - so before merging this, make sure that the unnecessary spacing is removed, like in [this commit](https://github.com/umbraco/UmbracoDocs/pull/5241/commits/ee4c4de1f4a5fca28677fee5638709ff075923c7) from before._

------

### CODE REVIEW

File: 12/umbraco-cms/reference/content-delivery-api/extension-points/custom-property-editors-support.md [hunk 1]
```diff
@@ -0,0 +1,227 @@
...
+
+## Implementation
+
+To implement the custom data values for your output in the Delivery API, we need to opt-in by implementing the `IDeliveryApiPropertyValueConverter` interface.
+
+The code example below showcases the implementation of a property editor called _"My.Custom.Picker"_. Our focus will be on the methods provided by the `IDeliveryApiPropertyValueConverter`, as they are responsible for customizing the Delivery API response. Towards the end of the file, you will find the response models that we will be using. The `IsConverter()` and `GetPropertyValueType()` methods are inherited from the `PropertyValueConverterBase` class, which is covered in the [Property Value Converters](http://localhost:5000/s/OdQETpqkO0Kcv8KMquKL/extending/property-editors/property-value-converters) article.
```
Comment (elit0451) [inline 12/umbraco-cms/reference/content-delivery-api/extension-points/custom-property-editors-support.md:21]: There are some weird links to the https://docs.umbraco.com/umbraco-cms/tutorials/creating-a-property-editor https://docs.umbraco.com/umbraco-cms/extending/property-editors/property-value-converters/ articles. But I couldn't find them in the 12.latest space, so I had to link them through v11 folder which resulted in this. Some help from the Docs team is appreciated here!

P.S. I have no idea why it says that a sentence in ln.21 is longer than 24 words. I went though each of them with a word count

File: 12/umbraco-cms/reference/content-delivery-api/extension-points/custom-property-editors-support.md [hunk 2]
```diff
@@ -0,0 +1,227 @@
...
+description: >-
+  Discover how to customize the Content Delivery API's response for your custom
+  property editors.
+---
+
+# Custom property editors support
+
+Out of the box, the Delivery API supports custom property editors, ensuring they are rendered alongside the built-in ones in Umbraco. However, if the output generated by your property editor isn't optimal for a headless context, you have the ability to customize the API response. This customization won't impact the Razor rendering, allowing you to tailor the Content Delivery API response according to your specific requirements.
+
+In this article, we'll look into how you can work with the IDeliveryApiPropertyValueConverter interface and implement custom [output expansion](../#output-expansion) for your custom property editors.
```
Comment (kjac) [inline 12/umbraco-cms/reference/content-delivery-api/extension-points/custom-property-editors-support.md:11]:
```suggestion
In this article, we'll look into how you can work with the `IDeliveryApiPropertyValueConverter` interface and implement custom [output expansion](../#output-expansion) for your custom property editors.
```

File: 12/umbraco-cms/reference/content-delivery-api/extension-points/custom-property-editors-support.md [hunk 3]
```diff
@@ -0,0 +1,227 @@
...
+
+## Prerequisite
+
+For the purpose of this example, it is assumed that you have created a custom property editor that stores the key of a content item. We will not dive into the details of creating a custom property editor and the necessary implementation steps within Umbraco for this article. If you need guidance on that, please refer to the [Creating a Property Editor](http://localhost:5000/s/OdQETpqkO0Kcv8KMquKL/tutorials/creating-a-property-editor) and [Property Value Converters](http://localhost:5000/s/OdQETpqkO0Kcv8KMquKL/extending/property-editors/property-value-converters) articles.
```
Comment (kjac) [inline 12/umbraco-cms/reference/content-delivery-api/extension-points/custom-property-editors-support.md:15]:
```suggestion
The examples in this article revolve around the fictional property editor My.Custom.Picker. This property editor stores the key of a single content item, and is backed by a property value converter.

We will not dive into the details of creating a custom property editor for Umbraco in this article. If you need guidance on that, please refer to the [Creating a Property Editor](http://localhost:5000/s/OdQETpqkO0Kcv8KMquKL/tutorials/creating-a-property-editor) and [Property Value Converters](http://localhost:5000/s/OdQETpqkO0Kcv8KMquKL/extending/property-editors/property-value-converters) articles.
```

File: 12/umbraco-cms/reference/content-delivery-api/extension-points/custom-property-editors-support.md [hunk 4]
```diff
@@ -0,0 +1,227 @@
...
+
+    public IApiContentRoute? Route { get; set; }
+}
+```
+{% endcode %}
+
+The Implementation of the `IDeliveryApiPropertyValueConverter` interface can be found in the following methods:
+
+* `GetDeliveryApiPropertyCacheLevel()`: This method specifies the cache level used for our property representation in the Delivery API response.
+* `GetDeliveryApiPropertyValueType()`: This method defines the value type of the custom property for the Delivery API response. The value type for the API may differ from the type used for Razor rendering _(as specified by `GetPropertyValueType()`)_.
```
Comment (kjac) [inline 12/umbraco-cms/reference/content-delivery-api/extension-points/custom-property-editors-support.md:111]:
```suggestion
* `GetDeliveryApiPropertyValueType()`: This method defines the value type of the custom property output for the Delivery API response.
```

Everything can differ from the Razor rendering - that's the whole point. Is there any reason to single out the value type here?

File: 12/umbraco-cms/reference/content-delivery-api/extension-points/custom-property-editors-support.md [hunk 5]
```diff
@@ -0,0 +1,227 @@
...
+  * In the given example, having only the `Guid` as the value is sufficient for the Razor rendering. Since Razor provides full access to the published content within its rendering context, which is not the case with the Content Delivery API.
```
Comment (kjac) [inline 12/umbraco-cms/reference/content-delivery-api/extension-points/custom-property-editors-support.md:112]:
```suggestion
```

Suggest to move this below (and rephrase) - see comment below.

File: 12/umbraco-cms/reference/content-delivery-api/extension-points/custom-property-editors-support.md [hunk 6]
```diff
@@ -0,0 +1,227 @@
...
+* `ConvertIntermediateToDeliveryApiObject()`: This method converts the value from the property editor to the desired custom object in a headless context.
+
```
Comment (kjac) [inline 12/umbraco-cms/reference/content-delivery-api/extension-points/custom-property-editors-support.md:114]:
```suggestion

In the given example, the content key (`Guid` value) is used when rendering with Razor. This is sufficient, because Razor provides full access to the published content within the rendering context. However, in a headless context we do not have the same access. To prevent subsequent round-trips to the server, we create a richer output model specifically for the Delivery API.

```

File: 12/umbraco-cms/reference/content-delivery-api/extension-points/extension-api-for-querying.md [hunk 7]
```diff
@@ -0,0 +1,308 @@
...
+description: >-
+  Learn how to extend the Content Delivery API with custom selecting, filtering,
+  and sorting options for the multi-item-based endpoint.
+---
+
+# Extension API for querying
+
+The Delivery API allows you to retrieve multiple items by utilizing the `/umbraco/delivery/api/v1/content` endpoint. With the built-in query parameters, you have the flexibility to get any number of content nodes based on your needs. For a comprehensive list of supported query options, please refer to the [Endpoints](../#endpoints) section.
+
+For the query endpoint, we have created a new Examine index (**DeliveryApiContentIndex**) that facilitates fast retrieval of the desired content. This index ensures quick indexing and searching of data, with the possibility for future extensions. In this article, we'll explore creating custom selecting, filtering, and sorting options to enhance the querying capabilities of the Delivery API.
```
Comment (kjac) [inline 12/umbraco-cms/reference/content-delivery-api/extension-points/extension-api-for-querying.md:11]: It looks like you've made all references to the index bold. Did you mean to? 

It seems a little pronounced when the markdown is rendered. Perhaps italic would be a better choice?

File: 12/umbraco-cms/reference/content-delivery-api/extension-points/extension-api-for-querying.md [hunk 8]
```diff
@@ -0,0 +1,308 @@
...
+
+## Query options
+
+Let's take a look at an example of using the query endpoint with query parameters for `fetch`, `filter`, and `sort`. For example, a request to the Delivery API may look like this:
```
Comment (kjac) [inline 12/umbraco-cms/reference/content-delivery-api/extension-points/extension-api-for-querying.md:15]:
```suggestion
Let's take a look at an example of using the query endpoint with query parameters for `fetch`, `filter`, and `sort`. A request might look like this:
```

File: 12/umbraco-cms/reference/content-delivery-api/extension-points/extension-api-for-querying.md [hunk 9]
```diff
@@ -0,0 +1,308 @@
...
+
+```http
+GET /umbraco/delivery/api/v1/content?fetch=xxx&filter=yyy&filter=zzz&sort=aaa&sort=bbb
+```
+
+The placeholders in the example stand in for the values that each query option checks to determine which one can handle the given query.&#x20;
```
Comment (kjac) [inline 12/umbraco-cms/reference/content-delivery-api/extension-points/extension-api-for-querying.md:21]: I don't entirely understand this sentence. Can you rephrase it?

File: 12/umbraco-cms/reference/content-delivery-api/extension-points/extension-api-for-querying.md [hunk 10]
```diff
@@ -0,0 +1,308 @@
...
+{% endhint %}
+
+
+
+The implementation of each querying option consists of a class for indexing the data into the **DeliveryApiContentIndex** and another one for handling the query. By implementing the `IContentIndexHandler` interface, you can control how your relevant data is indexed and made available for querying through our index. And you can customize the querying behaviour to suit your needs by implementing the `ISelectorHandler`, `IFilterHandler`, and `ISortHandler` interfaces.
+
+In the following sections, we will explore the implementation details of creating custom querying functionality for the Delivery API.
+
+## Custom selector
+
```
Comment (kjac) [inline 12/umbraco-cms/reference/content-delivery-api/extension-points/extension-api-for-querying.md:34]:
```suggestion

Selectors handle the `fetch` part of a query.

```

File: 12/umbraco-cms/reference/content-delivery-api/extension-points/extension-api-for-querying.md [hunk 11]
```diff
@@ -0,0 +1,308 @@
...
+To showcase how a custom selector is implemented consider a simple site structure with a few blog posts. Each post links to an author from a predefined list. Authors, on the other hand, can be marked as _'Featured'_ using a toggle on their Document Type, granting them additional visibility and recognition. We will use this marker as part of the indexing implementation for our selector option.
```
Comment (kjac) [inline 12/umbraco-cms/reference/content-delivery-api/extension-points/extension-api-for-querying.md:35]:
```suggestion
To showcase how to build a custom selector,  consider a site structure with a few blog posts. A post is linked to an author, which is another content item.

Authors can be marked as _'Featured'_ using a toggle on their Document Type, granting them additional visibility and recognition. We will use this marker as part of the indexing implementation for our selector option.
```

File: 12/umbraco-cms/reference/content-delivery-api/extension-points/extension-api-for-querying.md [hunk 12]
```diff
@@ -0,0 +1,308 @@
...
+                FieldName = FieldName,
+                FieldType = FieldType.StringRaw,
+                VariesByCulture = false
+            }
+        };
+}
+```
+{% endcode %}
+
+The `AuthorSelector` class implements the `ISelectorHandler` and `IContentIndexHandler` interfaces. The `ISelectorHandler` allows for handling the custom selector in API queries through the `CanHandle()` and `BuildSelectorOption()` methods. The implementation of `CanHandle()` determines if the given `fetch` query corresponds to the `"featuredAuthors"` value. The `BuildSelectorOption()` method constructs the selector option to search for authors with a positive value (e.g., `"y"`) in a `"featured"` index field. This implementation is very customizable, giving you the flexibility to define your own logic.&#x20;
```
Comment (kjac) [inline 12/umbraco-cms/reference/content-delivery-api/extension-points/extension-api-for-querying.md:96]:
```suggestion
The `AuthorSelector` class implements the `ISelectorHandler` and `IContentIndexHandler` interfaces.

`ISelectorHandler` allows for handling the `fetch` value in API queries through the `CanHandle()` and `BuildSelectorOption()` methods.

- `CanHandle()` determines if the given `fetch` query corresponds to the `"featuredAuthors"` value. 
- `BuildSelectorOption()` method constructs the selector option to search for authors with a positive value (e.g., `"y"`) in a `"featured"` index field.
```

File: 12/umbraco-cms/reference/content-delivery-api/extension-points/extension-api-for-querying.md [hunk 13]
```diff
@@ -0,0 +1,308 @@
...
+
+For indexing, both the `GetFieldValues()` and `GetFields()` methods play a crucial role in defining how the data should be indexed and made searchable. The `GetFieldValues()` method is responsible for retrieving the value of the _'Featured'_ field of content items of type _'author'_. Then creating an `IndexFieldValue` with the appropriate field value (`"y"` for featured, `"n"` otherwise) which will be used when populating the index. `GetFields()` defines the `"featured"` field in the **DeliveryApiContentIndex**, allowing for efficient and accurate searching.
```
Comment (kjac) [inline 12/umbraco-cms/reference/content-delivery-api/extension-points/extension-api-for-querying.md:98]:
```suggestion
The `GetFields()` and `GetFieldValues()` methods each play a role in defining how the data should be indexed and made searchable.

- `GetFields() defines` the behavior of fields that are added to the index. In this example, the "featured" field is added as a "raw" string for efficient and accurate searching.
- `GetFieldValues()` is responsible for retrieving the values of the defined index fields. In this case, the 'Featured' field of content items of type 'author'. It creates an `IndexFieldValue` with the appropriate field value (`"y"` for featured, `"n"` otherwise), which will be added to the index.
```

File: 12/umbraco-cms/reference/content-delivery-api/extension-points/extension-api-for-querying.md [hunk 14]
```diff
@@ -0,0 +1,308 @@
...
+
+Since our custom query option modifies the index structure, we will need to rebuild the **DeliveryApiContentIndex**. You can find it by navigating to the "Examine Management" dashboard in the "Settings" section. After that, we can make a request to the Delivery API query endpoint as follows:
```
Comment (kjac) [inline 12/umbraco-cms/reference/content-delivery-api/extension-points/extension-api-for-querying.md:100]:
```suggestion
Since our custom query option modifies the index structure, we will need to rebuild the **DeliveryApiContentIndex**. You can find it by navigating to the "Examine Management" dashboard in the "Settings" section. Once rebuilt, we can make a request to the Delivery API query endpoint as follows:
```

File: 12/umbraco-cms/reference/content-delivery-api/extension-points/extension-api-for-querying.md [hunk 15]
```diff
@@ -0,0 +1,308 @@
...
+    "total": 3,
+    "items": [
+        ...
+    ]
+}
+```
+
+## Custom filter
+
+Staying within the topic of blog posts and their authors, we can introduce a custom filter which will narrow down our content retrieval. With this custom filter, we can fetch specific blog posts associated with one or more authors based on our filtering query. We can specify the author(s) by providing their content item's key (GUID) as the value of the `author:` filter option. Multiple authors can be included by listing their GUIDs as comma-separated-values, like:
```
Comment (kjac) [inline 12/umbraco-cms/reference/content-delivery-api/extension-points/extension-api-for-querying.md:121]:
```suggestion
Filters handle the `filter` part of a query. Staying within the topic of blog posts and their authors, we will create a custom filter to find posts by author

Remember that authors are content items like posts. The filter allows specifying the desired author(s) by their key (`Guid`) in an `author:` filter option. Multiple authors can be included by listing their keys as comma-separated-values, like:
```

File: 12/umbraco-cms/reference/content-delivery-api/extension-points/extension-api-for-querying.md [hunk 16]
```diff
@@ -0,0 +1,308 @@
...
+
+**Request**
+
+```http
+GET /umbraco/delivery/api/v1/content?filter=author:7c630f15-8d93-4980-a0fc-027314dc827a,75380b4f-6d6e-47a1-9222-975cdfb2ac5f
+```
+
+The response will include the blog posts associated with the provided authors' IDs, enabling us to get only the relevant results from the API.
```
Comment (kjac) [inline 12/umbraco-cms/reference/content-delivery-api/extension-points/extension-api-for-querying.md:129]:
```suggestion
The response will include the blog posts associated with the provided authors, enabling us to retrieve only the relevant results from the API.
```

File: 12/umbraco-cms/reference/content-delivery-api/extension-points/extension-api-for-querying.md [hunk 17]
```diff
@@ -0,0 +1,308 @@
...
+    "total": 4,
+    "items": [
+        ...
+    ]
+}
+```
+
+Our filter implementation follows a similar structure to the custom selector we discussed earlier. We continue to utilize the `IContentIndexHandler` interface, but this time we introduce the `IFilterHandler`. This combination gives us flexibility and control over the filtering behaviour.
+
+The underlying idea remains consistent: we store the author's `Guid` in a new `"authorId"` field within the index. Consequently, we will need to rebuild the index to reflect the changes.
```
Comment (kjac) [inline 12/umbraco-cms/reference/content-delivery-api/extension-points/extension-api-for-querying.md:144]:
```suggestion
The procedure remains the same: We store and query the author key in a new `"authorId"` field within the index. Consequently, we will need to rebuild the index to reflect the changes.
```

File: 12/umbraco-cms/reference/content-delivery-api/extension-points/extension-api-for-querying.md [hunk 18]
```diff
@@ -0,0 +1,308 @@
...
+            FieldName = FieldName,
+            FieldType = FieldType.StringRaw,
+            VariesByCulture = false
+        }
+    };
+}
+```
+{% endcode %}
+
+This filter performs an exact match for each specified GUID against the index. The Delivery API also supports "Contains" filters, which can be used, for instance, to create an author filter based on names instead of IDs. In this case, the `BuildFilterOption()` method should use the `FilterOperation.Contains` operator, and the `GetFields()` implementation should specify `FieldType.StringAnalyzed` for the returned `IndexField`.
```
Comment (kjac) [inline 12/umbraco-cms/reference/content-delivery-api/extension-points/extension-api-for-querying.md:213]:
```suggestion
The principal difference from the selector is that the filter implements `BuildFilterOption()` instead of `BuildSelectorOption()`. Here, the filter performs an exact match for any specified `Guid` in the query. Efficiently this makes the filter perform an `OR` operation against the index. 

Since we need to perform an exact match, the index field (`authorId`) is once again defined as a "raw" string. Other options include "analyzed" and "sortable" strings. These support "contains" searches and alpha-numeric sorting, respectively.
```

File: 12/umbraco-cms/reference/content-delivery-api/extension-points/extension-api-for-querying.md [hunk 19]
```diff
@@ -0,0 +1,308 @@
...
+
+## Custom sort
+
+Finally, adding a new custom sorting option can enable us to sort our blog posts based on a custom `"publishDate"` Date Picker property. This allows us to arrange the posts in ascending or descending order based on their publication dates.
```
Comment (kjac) [inline 12/umbraco-cms/reference/content-delivery-api/extension-points/extension-api-for-querying.md:217]:
```suggestion
Finally, we can also add custom handling of the `sort` part of the query. 
We'll add a custom sort handler that allows us to sort blog posts based on a custom `"publishDate"` Date Picker property. The implementation will allow for sorting the posts in ascending or descending order.
```

File: 12/umbraco-cms/reference/content-delivery-api/extension-points/extension-api-for-querying.md [hunk 20]
```diff
@@ -0,0 +1,308 @@
...
+
+{% hint style="info" %}
+This sorting should only be used with content nodes that have a published date to ensure accurate results.
```
Comment (kjac) [inline 12/umbraco-cms/reference/content-delivery-api/extension-points/extension-api-for-querying.md:220]:
```suggestion
This sorting should only be used with query results that have a published date to ensure accurate results.
```

File: 12/umbraco-cms/reference/content-delivery-api/extension-points/extension-api-for-querying.md [hunk 21]
```diff
@@ -0,0 +1,308 @@
...
+                FieldName = FieldName,
+                FieldType = FieldType.Date,
+                VariesByCulture = false
+            }
+        };
+}
+```
+{% endcode %}
+
+The provided implementation follows the familiar structure defined by the `IContentIndexHandler` and `ISortHandler` interfaces. And another point to highlight is that we store the `"publishDate"` value in a `Date` format within the index, which requires regenerating the data in the **DeliveryApiContentIndex**.
```
Comment (kjac) [inline 12/umbraco-cms/reference/content-delivery-api/extension-points/extension-api-for-querying.md:289]:
```suggestion
The implementation follows the same structure as the other examples, defined by the `IContentIndexHandler` and `ISortHandler` interfaces.

One point to highlight is that we store the `"publishDate"` value as a "date" field in index, which allows for correct date sorting.

Once more, when adding fields to the index, we need to rebuild it to reflect the changes.
```

File: 12/umbraco-cms/reference/content-delivery-api/extension-points/extension-api-for-querying.md [hunk 22]
```diff
@@ -0,0 +1,308 @@
...
+
+In the following example request, we also apply the author filter to retrieve only `"blogpost"` content nodes, which we know have the `"publishDate"` field. This helps to ensure that our sorting yields meaningful and relevant results.
```
Comment (kjac) [inline 12/umbraco-cms/reference/content-delivery-api/extension-points/extension-api-for-querying.md:291]: To follow the rest of the article, I think you should lead the "Custom sort" section with the request and the response, before diving into the implementation.

File: 12/umbraco-cms/reference/content-delivery-api/extension-points/custom-property-editors-support.md (hunk 4 above)
Comment (elit0451) [reply 12/umbraco-cms/reference/content-delivery-api/extension-points/custom-property-editors-support.md:111]: You are absolutely right! There wasn't really any particular reason for it, just thought it is an easy-to-relate example 😁

File: 12/umbraco-cms/reference/content-delivery-api/extension-points/custom-property-editors-support.md (hunk 6 above)
Comment (elit0451) [reply 12/umbraco-cms/reference/content-delivery-api/extension-points/custom-property-editors-support.md:114]: Makes a good of sense!

File: 12/umbraco-cms/reference/content-delivery-api/extension-points/extension-api-for-querying.md (hunk 7 above)
Comment (elit0451) [reply 12/umbraco-cms/reference/content-delivery-api/extension-points/extension-api-for-querying.md:11]: It is only because this is how we started referencing it in the getting started article, but I can change that 😌

File: 12/umbraco-cms/reference/content-delivery-api/extension-points/extension-api-for-querying.md (hunk 22 above)
Comment (elit0451) [reply 12/umbraco-cms/reference/content-delivery-api/extension-points/extension-api-for-querying.md:291]: There isn't really a pattern for when the example Request and Response should come - in the Selector they are after the implementation and in the Filter section they are before. So I think it is fine to keep the Sort structure as it is.

And good point for repeating the relevant results part

**Review by kjac — APPROVED** [review]
Brilliant stuff 🚀 

I haven't looked at the updated files, but the newly added files (the articles and the readme) look great 🎉

File: 12/umbraco-cms/reference/content-delivery-api/extension-points/README.md [hunk 23]
```diff
@@ -0,0 +1,12 @@
+---
+description: A guide to extending the Content Delivery API.
+---
+
+# Extension points
+
+The Delivery API has been designed with extensibility in mind, offering multiple extension points that provide greater flexibility and customization options. This article explores these extension points, allowing you to tailor the API's behaviour and expand its capabilities to meet your specific requirements.
```
Comment (sofietoft) [inline 12/umbraco-cms/reference/content-delivery-api/extension-points/README.md:7]: "This article explores these extension points..."

Is something missing from this article?
It only list links, and doesn't seem to really explore anything? I might be wrong 🤔

File: 12/umbraco-cms/reference/content-delivery-api/extension-points/custom-property-editors-support.md [hunk 24]
```diff
@@ -0,0 +1,242 @@
...
+In this article, we'll look into how you can work with the `IDeliveryApiPropertyValueConverter` interface and implement custom [output expansion](../#output-expansion) for your custom property editors.
+
+## Prerequisite
+
+The examples in this article revolve around the fictional `My.Custom.Picker` property editor. This property editor stores the key of a single content item and is backed by a property value converter.
+
+We will not dive into the details of creating a custom property editor for Umbraco in this article. If you need guidance on that, please refer to the [Creating a Property Editor](http://localhost:5000/s/OdQETpqkO0Kcv8KMquKL/tutorials/creating-a-property-editor) and [Property Value Converters](http://localhost:5000/s/OdQETpqkO0Kcv8KMquKL/extending/property-editors/property-value-converters) articles.
```
Comment (sofietoft) [inline 12/umbraco-cms/reference/content-delivery-api/extension-points/custom-property-editors-support.md:17]:
```suggestion
We will not dive into the details of creating a custom property editor for Umbraco in this article. If you need guidance on that, please refer to the [Creating a Property Editor](https://docs.umbraco.com/umbraco-cms/tutorials/creating-a-property-editor) and [Property Value Converters](https://docs.umbraco.com/umbraco-cms/extending/property-editors/property-value-converters) articles.
```

File: 12/umbraco-cms/reference/content-delivery-api/extension-points/custom-property-editors-support.md [hunk 25]
```diff
@@ -0,0 +1,242 @@
...
+
+## Implementation
+
+To customize the output of a property value editor in the Delivery API, we need to opt-in by implementing the `IDeliveryApiPropertyValueConverter` interface.
+
+The code example below showcases the implementation of this interface in the property value converter for `My.Custom.Picker`. Our focus will be on the methods provided by the `IDeliveryApiPropertyValueConverter`, as they are responsible for customizing the Delivery API response.
+
+Towards the end of the example, you will find the response models that we will be using.
+
+The `IsConverter()` and `GetPropertyValueType()` methods are inherited from the `PropertyValueConverterBase` class, which is covered in the [Property Value Converters](http://localhost:5000/s/OdQETpqkO0Kcv8KMquKL/extending/property-editors/property-value-converters) article.
```
Comment (sofietoft) [inline 12/umbraco-cms/reference/content-delivery-api/extension-points/custom-property-editors-support.md:27]:
```suggestion
The `IsConverter()` and `GetPropertyValueType()` methods are inherited from the `PropertyValueConverterBase` class, which is covered in the [Property Value Converters](https://docs.umbraco.com/umbraco-cms/extending/property-editors/property-value-converters) article.
```

File: 12/umbraco-cms/reference/content-delivery-api/extension-points/extension-api-for-querying.md [hunk 26]
```diff
@@ -0,0 +1,331 @@
...
+description: >-
+  Learn how to extend the Content Delivery API with custom selecting, filtering,
+  and sorting options for the multi-item-based endpoint.
+---
+
+# Extension API for querying
+
+The Delivery API allows you to retrieve multiple items by utilizing the `/umbraco/delivery/api/v1/content` endpoint. With the built-in query parameters, you have the flexibility to get any number of content nodes based on your needs. For a comprehensive list of supported query options, please refer to the [Endpoints](../#endpoints) section.
+
+For the query endpoint, we have created a new Examine index (_DeliveryApiContentIndex_) that facilitates fast retrieval of the desired content. This index ensures quick indexing and searching of data, with the possibility for future extensions. In this article, we'll explore creating custom selecting, filtering, and sorting options to enhance the querying capabilities of the Delivery API.
```
Comment (sofietoft) [inline 12/umbraco-cms/reference/content-delivery-api/extension-points/extension-api-for-querying.md:11]:
```suggestion
For the query endpoint, we have created a new Examine index (_DeliveryApiContentIndex_) that facilitates fast retrieval of the desired content. This index ensures quick indexing and searching of data, with the possibility for future extensions.

In this article, we'll explore creating custom selecting, filtering, and sorting options to enhance the querying capabilities of the Delivery API.
```

File: 12/umbraco-cms/reference/content-delivery-api/extension-points/extension-api-for-querying.md [hunk 27]
```diff
@@ -0,0 +1,331 @@
...
+}
+```
+{% endcode %}
+
+The `AuthorSelector` class implements the `ISelectorHandler` and `IContentIndexHandler` interfaces.
+
+`ISelectorHandler` allows for handling the `fetch` value in API queries through the `CanHandle()` and `BuildSelectorOption()` methods.
+
+* `CanHandle()` determines if the given `fetch` query corresponds to the `"featuredAuthors"` value. 
+* `BuildSelectorOption()` method constructs the selector option to search for authors with a positive value (for example, `"y"`) in a `"featured"` index field.
```
Comment (sofietoft) [inline 12/umbraco-cms/reference/content-delivery-api/extension-points/extension-api-for-querying.md:106]: "methods" seems a little redundant here, so I've suggested to take it out.

```suggestion
* `BuildSelectorOption()` constructs the selector option to search for authors with a positive value (for example, `"y"`) in a `"featured"` index field.
```

**Review by sofietoft — COMMENTED** [review]
Great PR @elit0451 !
Really well written and the article has a nice flow to them 💪 

As I'm adding this review, I haven't yet previewed the articles on GitBook - I'll do that once this is submitted.
I've added a couple of suggestions, and also have these general comments:

Is the Extension points article necessary? Seems a little redundant to have this "middle article" that's not really doing anything.
Seems like the section in this article could be added to the main "Content Delivery API" article?

About the broken "localhost" links.
The property editor articles aren't in the 12 space yet, as they do not need to be updated for Umbraco 12. We'll merge it all together once the version is released.
Until then, let's link to the v11 version of these articles - I've suggested the change!

File: 12/umbraco-cms/reference/content-delivery-api/extension-points/README.md (hunk 23 above)
Comment (elit0451) [reply 12/umbraco-cms/reference/content-delivery-api/extension-points/README.md:7]: You are right! Maybe it doesn't make sense to be here as you suggested 🙂

**Review by sofietoft — APPROVED** [review]
Looks great @elit0451 ! 💪 

Ping me when you're happy with how it looks on GitBookk as well, and I'll get it merged!
